_episodes/05-preprocessor.md

```diff
@@ -0,0 +1,287 @@
+
+> ## Exercise: Adding more preprocessor steps
+>
+> Edit the [example recipe](LINK to episode #4) to change the variable in ?thetao? and to add preprocessors which take the average over the latitude and longitude dimensions and the average over the depth. And then run the recipe.
+>
+>> ## Solution
+>> 
+>>~~~YAML
+>> preprocessors:
+>>   prep_timeseries:
+>>     annual_statistics:
+>>       operator: mean
+>>     area_statistics:
+>>       operator: mean
+>>     depth_integration:
+>>~~~
+>>{: .source}
+>{: .solution}
+{: .challenge}
+
+## Using different preprocessors for different variables
+
+You could define different preprocessors with several preprocessor sections (setting different preprocessor names). In the variable section you call the specific preprocessor which should be applied.
+
+> ## Example
+>~~~YAML
+> preprocessors:
+>   prep_timeseries_1:
+>     annual_statistics:
+>       Operator: mean
+>   prep_timeseries_2:
+>     annual_statistics:
+>       operator: mean
+>     area_statistics:
+>       operator: mean
+>     Depth_integration:
+> ---
+> diagnostics:
+>   # --------------------------------------------------      
+>   # Time series diagnostics
+>   # --------------------------------------------------
+>   diag_timeseries_temperature_1:
+>     description: simple_time_series>
+>     variables:
+>         timeseries_variable:
+>         short_name: thetaoga
+>         preprocessor: prep_timeseries_1
+>     scripts:
+>          timeseries_diag:
+>         script: ocean/diagnostic_timeseries.py
+> 
+>   diag_timeseries_temperature_2:
+>     description: simple_time_series
+>     variables:
+>         timeseries_variable:
+>         short_name: thetao
+>         preprocessor: prep_timeseries_2
+>     scripts:
+>          timeseries_diag:
+>         script: ocean/diagnostic_timeseries.py
+>~~~
+>{: .source}
+{: .solution}
+
+>## Challenge : How to write a recipe with multiple preprocessors
+> We now know that a recipe can have more than one diagnostic, variable or preprocessor. As we saw in the examples so far, we can group preprocessors with a single user defined name and can have more than one such preprocessor group in the recipe as well. Write two different preprocessors - one to regrid the data to a 1x1 resolution and the second preprocessor to mask out sea and ice grid cells before regridding to the saem resolution. In the second case, ensure you perform the masking first before regridding (hint: custom order your operations). Now, use the two preprocessors in different diagnostics within the same recipe. You may use any variable(s) of your choice. Once you succeed, try to add new datasets to the same recipe. Placeholders for the different components are provided below:
+>
+>> ## Recipe
+>>
+>>~~~YAML
+>>
+>> datasets:
+>>   - {dataset: UKESM1-0-LL, project: CMIP6, exp: historical,
```

Comment:
I think we should make sure that all episodes refer to the same project. Here you refer to CMIP6 data, but I think in episode 3 we refer to CMIP5 right now...

Reply:
Let's try to align those sort of things in a next pass over the entire tutorial, once most of the content is merged.